src/System.Management.Automation/engine/Attributes.cs

```diff
@@ -943,43 +970,100 @@ protected override void ValidateElement(object element)
                         Metadata.ValidateNotNullFailure);
             }
 
-            var o = element as PSObject;
-            if (o != null)
+            if(_rangeKind.HasValue)
             {
-                element = o.BaseObject;
+                // Using dynamic to handle different type of numbers such as
+                // int, double, decimal and float. 
+                dynamic val = element;
+                switch (_rangeKind)
+                {
+                    case ValidateRangeKind.Positive:
+                        if(val.CompareTo(0) <= 0)
```

Comment:
If the argument does not implement `IComparable`, the error message will not be good.